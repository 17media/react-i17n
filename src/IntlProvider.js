import React, { PureComponent } from 'react';
import IntlMessageFormat from 'intl-messageformat';
import Aux from './Aux';
import Intl from './Intl';
import { Provider } from './IntlContext';

class IntlProvider extends PureComponent {
  static defaultProps = {
    loadingTimeout: 500, // 500 ms of blank instead of fallback message
    WrappedComponent: null, // default to not render the wrapper element
  };

  constructor(props) {
    super(props);

    this.state = {
      isTimeout: false,
    };
  }

  componentDidMount() {
    this.setTimeout();
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      isTimeout: false,
    }, () => {
      this.setTimeout();
    });
  }

  setTimeout = () => {
    this.timeout = global.setTimeout(() => {
      this.setState({
        isTimeout: true,
      });
    }, this.props.loadingTimeout);
  }

  render() {
    const { locale, messages, WrappedComponent, children } = this.props;
    const intl = new Intl(locale, messages);
    const { isTimeout } = this.state;

    const context = {
      intl,
      isTimeout,
      WrappedComponent,
    };

    return (
      <Provider value={context}>
        {children}
      </Provider>
    );
  }
}

export default IntlProvider;
