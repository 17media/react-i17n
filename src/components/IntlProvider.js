import React from 'react';
import { Provider } from '../IntlContext';
import Intl from '../Intl';

class IntlProvider extends React.PureComponent {
  static getDerivedStateFromProps(props, state) {
    if (
      state.intl.locale !== props.locale ||
      state.intl.messages !== props.messages
    ) {
      return {
        intl: new Intl(props.locale, props.messages),
      };
    }

    return null;
  }

  constructor(props) {
    super(props);

    this.state = {
      intl: new Intl(props.locale, props.messages),
    };
  }

  render() {
    const { intl } = this.state;
    const { children } = this.props;

    return <Provider value={intl}>{children}</Provider>;
  }
}

export default IntlProvider;
