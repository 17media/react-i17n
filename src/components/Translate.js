import React, { createElement, Fragment } from 'react';
import { IntlConsumer } from '../IntlContext';

const Translate = ({ id, children, defaultMessage, values }) => (
  <IntlConsumer>
    {intl => {
      const message = intl.formatMessage(children || id, {
        defaultMessage,
        values,
      });

      if (typeof message === 'string') {
        return message;
      }

      return createElement(Fragment, {}, ...message);
    }}
  </IntlConsumer>
);

export default Translate;
