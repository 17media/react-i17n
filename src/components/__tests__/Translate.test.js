import React from 'react';
import { create } from 'react-test-renderer';
import IntlProvider from '../IntlProvider';
import Translate from '../Translate';

describe('<Translate />', () => {
  const locale = 'en';
  const defaultMessageID = 'id';

  const messages = {
    [locale]: {
      [defaultMessageID]: '',
    },
  };

  afterEach(() => {
    messages[locale][defaultMessageID] = '';
  });

  test('translate simple message', () => {
    const message = 'Hello World!';

    messages[locale][defaultMessageID] = message;

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Translate>{defaultMessageID}</Translate>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(message);
  });

  test('fallback to default message if the id is not found', () => {
    const defaultMessage = 'Hello World!';

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Translate defaultMessage={defaultMessage}>no such id</Translate>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(defaultMessage);
  });

  test('fallback to empty string if both the id and defaultMessage are not found', () => {
    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Translate>no such id</Translate>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe('');
  });

  test('translate message with values', () => {
    const message = 'Hello {name}, Welcome to {site}!';
    const values = {
      name: 'Kai Hao',
      site: '17live',
    };

    messages[locale][defaultMessageID] = message;

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Translate values={values}>{defaultMessageID}</Translate>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe('Hello Kai Hao, Welcome to 17live!');
  });

  test('translate message with react components as values', () => {
    const Link = props => <a {...props} />;

    const message = 'Hello {name}, Welcome to {site}!';
    const name = <b>Kai Hao</b>;
    const site = <Link href="https://17.live">17live</Link>;
    const values = {
      name,
      site,
    };

    messages[locale][defaultMessageID] = message;

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Translate values={values}>{defaultMessageID}</Translate>
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();
    expect(testRenderer.root.findByType(Link)).toBeTruthy();
  });

  test('take id attribute when children is not specified', () => {
    const message = 'Hello World!';

    messages[locale][defaultMessageID] = message;

    const testRenderer = create(
      <IntlProvider locale={locale} messages={messages}>
        <Translate id={defaultMessageID} />
      </IntlProvider>
    );

    expect(testRenderer.toJSON()).toBe(message);
  });
});
