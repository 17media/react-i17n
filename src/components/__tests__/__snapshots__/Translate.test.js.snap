// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Translate /> translate message with react components as values 1`] = `
Array [
  "Hello ",
  <b>
    Kai Hao
  </b>,
  ", Welcome to ",
  <a
    href="https://17.live"
  >
    17live
  </a>,
  "!",
]
`;
